src/UniswapV4MulticurveRehypeInitializerHook.sol

```diff
@@ -0,0 +1,612 @@
+pragma solidity ^0.8.24;
+
+import { BaseHook } from "@v4-periphery/utils/BaseHook.sol";
+import { IPoolManager } from "@v4-core/interfaces/IPoolManager.sol";
+import { Hooks } from "@v4-core/libraries/Hooks.sol";
+import { BalanceDelta, BalanceDeltaLibrary } from "@v4-core/types/BalanceDelta.sol";
+import { PoolKey } from "@v4-core/types/PoolKey.sol";
+import { PoolId } from "@v4-core/types/PoolId.sol";
+import { UniswapV4MulticurveInitializer } from "src/UniswapV4MulticurveInitializer.sol";
+import { TickMath } from "@v4-core/libraries/TickMath.sol";
+import { BeforeSwapDelta, BeforeSwapDeltaLibrary, toBeforeSwapDelta } from "@v4-core/types/BeforeSwapDelta.sol";
+import { StateLibrary } from "@v4-core/libraries/StateLibrary.sol";
+import { ProtocolFeeLibrary } from "@v4-core/libraries/ProtocolFeeLibrary.sol";
+import { Position } from "src/types/Position.sol";
+import { FullMath } from "@v4-core/libraries/FullMath.sol";
+import { SwapMath } from "@v4-core/libraries/SwapMath.sol";
+import { Currency, CurrencyLibrary } from "@v4-core/types/Currency.sol";
+import { Quoter } from "@quoter/Quoter.sol";
+import { LiquidityAmounts } from "@v4-periphery/libraries/LiquidityAmounts.sol";
+import { MigrationMath } from "src/libraries/MigrationMath.sol";
+import { IERC20 } from "lib/universal-router/lib/v4-periphery/lib/v4-core/lib/forge-std/src/interfaces/IERC20.sol";
+
+// goals
+// - create an empty full range LP position given tickSpacing
+// - save that position so that we can rehype it later
+// - when a swap happens, we should dynamically update the fee, and do a self swap to get 50/50
+// - when we get 50/50, we should add liquidity to the full range LP position
+// - after all is said and done, we should update the fee back to its original value
+
+/// @notice Thrown when the caller is not the Uniswap V4 Multicurve Initializer
+error OnlyInitializer();
+
+/**
+ * @notice Emitted when liquidity is modified
+ * @param key Key of the related pool
+ * @param params Parameters of the liquidity modification
+ */
+event ModifyLiquidity(PoolKey key, IPoolManager.ModifyLiquidityParams params);
+
+/**
+ * @notice Emitted when a Swap occurs
+ * @param sender Address calling the PoolManager
+ * @param poolKey Key of the related pool
+ * @param poolId Id of the related pool
+ * @param params Parameters of the swap
+ * @param amount0 Balance denominated in token0
+ * @param amount1 Balance denominated in token1
+ * @param hookData Data passed to the hook
+ */
+event Swap(
+    address indexed sender,
+    PoolKey indexed poolKey,
+    PoolId indexed poolId,
+    IPoolManager.SwapParams params,
+    int128 amount0,
+    int128 amount1,
+    bytes hookData
+);
+
+struct Fees {
+    uint128 fees0;
+    uint128 fees1;
+    uint24 customFee;
+}
+
+uint256 constant MAX_SWAP_FEE = SwapMath.MAX_SWAP_FEE;
+uint256 constant MAX_REBALANCE_ITERATIONS = 15;
+
+// TODO: factor in decimals haha 1e6 maybe for 18 decimals? idk haha
+uint128 constant EPSILON = 1e6;
```

Comment:
Just flagging this to point to: https://github.com/whetstoneresearch/doppler/pull/445#discussion_r2475529030

Deciding on a good EPSILON value is going to be very market specific. Not only does it depend on the decimals of the tokens, but also the value. It might even be necessary to have a different EPSILON for each token

If it's too high then it can be sandwiched, but if it's too low then we swap dust amounts excessively frequently